### PR TITLE
DOCSP-36993 Custom types

### PR DESCRIPTION
**Pull request by jordan-smith721**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-node/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-36993
Staging - https://preview-mongodbjordansmith721.gatsbyjs.io/pymongo/DOCSP-36993-custom-types/fundamentals/data-formats/custom-types/

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/fundamentals/data-formats/custom-types.txt [hunk 1]
```diff
@@ -1,233 +1,230 @@
...
 
 
-Since the purpose of the example is to demonstrate working with custom types,
-we'll need a custom data type to use. For this example, we will be working with
-the ``~decimal.Decimal`` type from Python's standard library. Since the
-BSON library's ``~bson.decimal128.Decimal128`` type (that implements
-the IEEE 754 decimal128 decimal-based floating-point numbering format) is
+This example uses the ``~decimal.Decimal`` type from Python's standard library. Since the
+BSON library's ``~bson.decimal128.Decimal128`` type--that implements
+the ``IEEE 754 decimal128 decimal-based floating-point`` numbering format--is
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:46]:
```suggestion
This example uses the ``~decimal.Decimal`` type from Python's standard library. The
BSON library's ``~bson.decimal128.Decimal128`` type, which implements
the IEEE 754 decimal128 decimal-based floating-point numbering format, is
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 2]
```diff
@@ -1,233 +1,230 @@
...
 distinct from Python's built-in ``~decimal.Decimal`` type, attempting
-to save an instance of ``Decimal`` with PyMongo, results in an
+to save an instance of ``Decimal`` with PyMongo, results in a
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:48]: Can't make a suggestion on it in GH, but I would change the end of line 48 to" "type. Trying"

Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:48]:
```suggestion
to save an instance of ``Decimal`` with PyMongo results in a
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 3]
```diff
@@ -1,233 +1,230 @@
...
+   >>> db.test.insert_one({"num": num})
+   Traceback (most recent call last):
+   ...
+   bson.errors.InvalidDocument: cannot encode object: Decimal('45.321'), of type: <class 'decimal.Decimal'>
 
 .. _custom-type-type-codec:
 
 The ``~bson.codec_options.TypeCodec`` Class
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:63]: don't think it's super important, but just for consistency

```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 4]
```diff
@@ -1,233 +1,230 @@
...
+   ...         custom type."""
+   ...         return value.to_decimal()
+   ...
+   >>> decimal_codec = DecimalCodec()
 
 .. _custom-type-type-registry:
 
 The ``~bson.codec_options.TypeRegistry`` Class
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:111]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 5]
```diff
@@ -236,44 +233,41 @@ Decoding ``~bson.binary.Binary`` Types
...
+   >>> # On Python 2.7.x
+   >>> newcoll = db.get_collection("new")
+   >>> doc = newcoll.find_one()
+   >>> type(doc["data"])
+   bson.binary.Binary
 
-As a consequence of this disparity, users must set the ``bson_type`` attribute
-on their ``~bson.codec_options.TypeDecoder`` classes differently,
+As a consequence of this disparity, you must set the ``bson_type`` attribute
+on you ``~bson.codec_options.TypeDecoder`` classes differently,
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:264]:
```suggestion
on your ``~bson.codec_options.TypeDecoder`` classes differently,
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 6]
```diff
@@ -282,157 +276,145 @@ The ``fallback_encoder`` Callable
...
-Because fallback encoders don't need to declare the types that they encode
-beforehand, they can be used to support interesting use-cases that cannot be
-serviced by ``TypeEncoder``. One such use-case is described in the next
-section.
-
+   Invoke fallback encoders *after* attempts to encode the particular value
+   with standard BSON encoders and any configured type encoders have failed.
+   Therefore, in a type registry configured with a type encoder and fallback
+   encoder that both target the same custom type, the behavior specified in
+   the type encoder prevails.
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:321]: The second sentence starts with "Therefore," which makes me think the first sentence is explaining something, but it's actually imperative ("Invoke fallback encoders...").

Does this suggestion match the meaning of the original?

```suggestion
   {+driver-short+} invokes fallback encoders *after* attempts to encode the particular value
   with standard BSON encoders and any configured type encoders have failed.
   Therefore, in a type registry configured with a type encoder and fallback
   encoder that both target the same custom type, the behavior specified in
   the type encoder prevails.
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 7]
```diff
@@ -282,157 +276,145 @@ The ``fallback_encoder`` Callable
...
 ^^^^^^^^^^^^^^^^^^^^^^
 
-In this example, we demonstrate how a fallback encoder can be used to save
-arbitrary objects to the database. We will use the the standard library's
-``pickle`` module to serialize the unknown types and so naturally, this
-approach only works for types that are picklable.
+This example describes how to use a fallback encoder to save
+arbitrary objects to the database. It uses the standard library's
+``pickle`` module to serialize the unknown types, and works only for types that
+are pickleable.
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:329]: had to google this, but the [pickle docs](https://docs.python.org/3/library/pickle.html#pickle.PickleBuffer) use "picklable", so i would go with that. (google also suggested "pickleball", which would be a fun addition to the docs.)

```suggestion
are picklable.
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 8]
```diff
@@ -282,157 +276,145 @@ The ``fallback_encoder`` Callable
...
-      )
-  )
+   codec_options = CodecOptions(
+       type_registry=TypeRegistry(
+           [PickledBinaryDecoder()], fallback_encoder=fallback_pickle_encoder
+       )
+   )
 
-We can now round trip our custom objects to MongoDB:
+You can now round trip your custom objects to MongoDB:
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:389]: another page i did used "round-trip" with a hyphen. i'm not sure which is correct, but when it's hyphenated, it's clearer to me that it's intended as a single verb

```suggestion
You can now round-trip your custom objects to MongoDB:
```

File: source/fundamentals/data-formats/custom-types.txt [hunk 9]
```diff
@@ -282,157 +276,145 @@ The ``fallback_encoder`` Callable
...
+  to all subtypes of the standard types.
+- Chaining type encoders is not supported. A custom type value, once
+  transformed by a codec's ``transform_python`` method, *must* result in a
+  type that is BSON-encodable by default, or that can be
+  transformed by the fallback encoder into something BSON-encodable. You
+  *cannot* transform it a second time by a different type codec.
+- The ``~pymongo.database.Database.command`` method does not apply the
+  user's TypeDecoders while decoding the command response document.
+- ``gridfs`` does not apply custom type encoding or decoding to any
+  documents received from, or returned to the user.
```
Comment (mongoKart) [inline source/fundamentals/data-formats/custom-types.txt:420]:
```suggestion
  documents received from, or returned to, the user.
```

**Review by mongoKart — CHANGES_REQUESTED** [review]
some small things to change but really a great job

File: source/fundamentals/data-formats/custom-types.txt (hunk 6 above)
Comment (jordan-smith721) [reply source/fundamentals/data-formats/custom-types.txt]: Good catch, your suggestion is what it should have been

File: source/fundamentals/data-formats/custom-types.txt (hunk 7 above)
Comment (jordan-smith721) [reply source/fundamentals/data-formats/custom-types.txt]: pickleball is preferred

**Review by mongoKart — APPROVED** [review]
Looks
Great
Thanks
Much

**Pull request merged**